cpp/test/sg/fil_test.cu

```diff
@@ -742,33 +743,39 @@ class TreeliteFilTest : public BaseFilTest {
     } else {
       int left          = root + 2 * (node - root) + 1;
       int right         = root + 2 * (node - root) + 2;
-      float threshold   = dense_node.thresh();
       bool default_left = dense_node.def_left();
-      switch (ps.op) {
-        case tl::Operator::kLT: break;
-        case tl::Operator::kLE:
-          // adjust the threshold
-          threshold = std::nextafterf(threshold, -std::numeric_limits<float>::infinity());
-          break;
-        case tl::Operator::kGT:
-          // adjust the threshold; left and right still need to be swapped
-          threshold = std::nextafterf(threshold, -std::numeric_limits<float>::infinity());
-        case tl::Operator::kGE:
-          // swap left and right
-          std::swap(left, right);
-          default_left = !default_left;
-          break;
-        default: ASSERT(false, "comparison operator must be <, >, <= or >=");
+      float threshold;
+      if (dense_node.is_categorical()) {
+        uint8_t byte = 0;
+        for (int category = 0; category <= cat_sets_h.max_matching[dense_node.fid()]; ++category) {
+          if (category % 8 == 0) byte = cat_sets_h.bits[dense_node.set() + category / 8];
+          if ((byte & 1 << category % 8) != 0) left_categories.push_back(category);
+        }
+      } else {
+        threshold = dense_node.thresh();
       }
       int left_key  = node_to_treelite(builder, pkey, root, left);
       int right_key = node_to_treelite(builder, pkey, root, right);
-      builder->SetNumericalTestNode(key,
-                                    dense_node.fid(),
-                                    ps.op,
-                                    tlf::Value::Create(threshold),
-                                    default_left,
-                                    left_key,
-                                    right_key);
+      if (!left_categories.empty() && dense_node.is_categorical()) {
+        std::swap(left_key, right_key);
+        default_left = !default_left;
+        builder->SetCategoricalTestNode(
+          key, dense_node.fid(), left_categories, default_left, left_key, right_key);
+      } else {
+        if (dense_node.is_categorical()) {
+          // treelite cannot handle empty category lists, so this threshold will always compare
+          // false
+          threshold = NAN;
+        }
+        adjust_threshold(&threshold, &left_key, &right_key, &default_left, ps.op, FIL_TO_TREELITE);
```

Comment:
Please avoid using `adjust_threshold` in tests. If necessary, you could implement a separate function for use in tests only.